Sources/TuistDependencies/Mappers/ExternalProjectsPlatformTreeShakerGraphMapper.swift

```diff
@@ -0,0 +1,128 @@
+ down from nodes in the graph that are closer to the root. This is a behaviour that we need to copy over to Tuist. In our case
+ the logic is executed at generation time.
+ */
+public final class ExternalProjectsPlatformTreeShakerGraphMapper: GraphMapping {
+    public init() {}
+
+    public func map(graph: Graph) async throws -> (Graph, [TuistCore.SideEffectDescriptor]) {
+        // If the project has no external dependencies we skip this.
+        if graph.projects.values.first(where: { $0.isExternal }) == nil {
+            return (graph, [])
+        }
+
+        var graph = graph
+
+        /**
+         One traverse gives us a dictionary that maps graph targets to the platforms they should support
+         */
+        let targetsToPlatformsDictionary = targetsToPlatformsDictionary(graph: graph)
+
+        graph.targets = Dictionary(uniqueKeysWithValues: graph.targets.map { projectPath, projectTargets in
+            let project = graph.projects[projectPath]!
+            let projectTargets = Dictionary(uniqueKeysWithValues: projectTargets.map { targetName, target in
+                let graphTarget = GraphTarget(path: projectPath, target: target, project: project)
+                var target = target
+
+                /**
+                 We only include the destinations whose platform is included in the list of the target supported platforms.
+                 */
+                if let targetFilteredPlatforms = targetsToPlatformsDictionary[graphTarget] {
+                    target.destinations = target.destinations.filter { destination in
+                        targetFilteredPlatforms.contains(destination.platform)
+                    }
+                }
+                return (targetName, target)
+            })
+            return (projectPath, projectTargets)
+        })
+        return (graph, [])
+    }
+
+    private func targetsToPlatformsDictionary(graph: Graph) -> [GraphTarget: Set<Platform>] {
+        let graphTraverser = GraphTraverser(graph: graph)
+        var platforms: [GraphTarget: Set<Platform>] = [:]
+
+        graphTraverser.allInternalTargets().forEach { target in
+            platforms[target] = target.target.supportedPlatforms
+            traverse(
+                target: target,
+                parentPlatforms: target.target.supportedPlatforms,
+                graphTraverser: graphTraverser,
+                platforms: &platforms
+            )
+        }
+
+        return platforms
+    }
+
+    private func traverse(
+        target: GraphTarget,
+        parentPlatforms: Set<Platform>,
+        graphTraverser: GraphTraverser,
+        platforms: inout [GraphTarget: Set<Platform>]
+    ) {
+        let dependencies = graphTraverser.directTargetDependenciesWithConditions(path: target.path, name: target.target.name)
+
+        dependencies.forEach { dependencyTarget, dependencyCondition in
```

Comment:
(already mentioned in DM) some of this logic seems duplicative of what we do within GraphTraverser `combinedCondition(to:from:)` [here](https://github.com/tuist/tuist/blob/f5512c0db26c2f952bd5b2a06a987de1a517a5af/Sources/TuistCore/Graph/GraphTraverser.swift#L660).  it would be nice if we could minimize redundant logic and refactor or extend the other method to allow for this use case if it does not already.  This was one of the more tricky bits of logic from the initial Multiplatform project.

Reply:
I tried to reuse the logic from `combinedCondition`, but I noticed the requirements are slightly different:
- When a condition is absent, we'd like to use the supported platforms of the parent as a reference and cascade that information down.
- We need to traverse through all the possible paths that take to a given external dependency from multiple (from) dependencies. 

Because of this, I decided to keep the logic separated instead of modifying the existing function because that'd make that function harder to reason about. I adjusted the function to only traverse from the ones that have an external dependency, as noted [here](https://github.com/tuist/tuist/pull/5674#discussion_r1424455534), and moved the logic to `GraphTraverser` to keep the mapper lean:

```swift
let externalTargetToPlatforms = graphTraverser.externalTargetSupportedPlatforms()
```